### PR TITLE
Add refund detail view

### PR DESCRIPTION
**Pull request by rlucioni**
Adds a refund detail view to Oscar which displays information about a single refund and allows a support team member to approve or deny that refund. Also provides UI elements on the refund list view which can be used to approve or deny a refund. Credit and revocation are coupled as far as the UI is concerned.

@clintonb and @jimabramson, could you please review this when you have the time? @Nickersoft, FYI.

@cptvitamin, this is the second half of the refund dashboard requiring an accessibility review. Whoever takes ownership of making your suggested changes (separate of getting this feature merged) will consider your comments on #100 in addition to any comments you may add here. Thank you for your help!

### CODE REVIEW

File: ecommerce/static/oscar/js/refund_list.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            headers: {'X-CSRFToken': $.cookie('csrftoken')}
+        }).success(function (data) {
+            $('tr[data-refund-id=' + refund_id + '] .refund-status').text(data.status);
+            addMessage('alert-success', 'icon-check-sign', 'Refund ' + refund_id + ' has been processed.');
+            $('[data-action=process-refund]').remove();
+        }).fail(function (jqXHR, textStatus, errorThrown) {
+            // NOTE (RFL): Uncertainty here. For an MVP, changing the displayed refund state on any error
+            // may be viable. Ideally, the displayed refund state would only change if the refund were
+            // to enter an error state. This would be easiest if the processing endpoint returned 200
+            // and a serialized refund, even when the refund has entered an error state during processing.
```
Comment (rlucioni) [inline ecommerce/static/oscar/js/refund_list.js]: @clintonb and @jimabramson, I'm interested in what you have to say about this.

What we can do here is dependent on the behavior of the refund processing endpoint. This generic error handler causes the refund's status to be displayed as "Error" if the Ajax request to the processing endpoint fails for any reason. I think it's enough for what we need now.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    @skip("Requires refund processing endpoint.")
+    @ddt.data(
+        (True, True),
+        (True, False),
+        (False, True),
+        (False, False),
+    )
+    @ddt.unpack
+    def test_processing_success(self, approve, detail):
```
Comment (clintonb) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: Consider renaming `detail` to something more informative (e.g. `test_detail_page`). You might also consider splitting these into two separate tests for the separate views.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        self.assert_alert_displayed('alert-success', 'Refund {} has been processed.'.format(self.refund.id))
+
+    @ddt.data(
+        (True, True),
+        (True, False),
+        (False, True),
+        (False, False),
+    )
+    @ddt.unpack
+    def test_processing_failure(self, approve, detail):
```
Comment (clintonb) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: Same as above regarding the argument name and splitting the tests.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        (REFUND.OPEN, False),
+        (REFUND.DENIED, True),
+        (REFUND.DENIED, False),
+        (REFUND.ERROR, True),
+        (REFUND.ERROR, False),
+        (REFUND.COMPLETE, True),
+        (REFUND.COMPLETE, False),
+    )
+    @ddt.unpack
+    def test_button_configurations(self, status, detail):
```
Comment (clintonb) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: Same as above regarding the argument name and splitting the tests.

File: ecommerce/static/oscar/js/refund_list.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+function addMessage(type, icon, msg) {
```
Comment (clintonb) [inline ecommerce/static/oscar/js/refund_list.js]: DRY: This is duplicated from `order_list.js`. Move it to a common file shared by both.

File: ecommerce/static/oscar/js/refund_list.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            // NOTE (RFL): Uncertainty here. For an MVP, changing the displayed refund state on any error
+            // may be viable. Ideally, the displayed refund state would only change if the refund were
+            // to enter an error state. This would be easiest if the processing endpoint returned 200
+            // and a serialized refund, even when the refund has entered an error state during processing.
+            $('tr[data-refund-id=' + refund_id + '] .refund-status').text('Error');
+
+            addMessage(
+                'alert-error',
+                'icon-exclamation-sign',
+                'Failed to process refund #' + refund_id + ': ' + errorThrown + '. Please try again or contact rogue-initiative@edx.org.'
```
Comment (clintonb) [inline ecommerce/static/oscar/js/refund_list.js]: Please try again, or contact the E-Commerce Development Team.

File: ecommerce/templates/oscar/dashboard/orders/order_list.html [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {% extends 'dashboard/layout.html' %}
 {% load compress %}
-{% load currency_filters %}
```
Comment (clintonb) [inline ecommerce/templates/oscar/dashboard/orders/order_list.html:3]: Unnecessary change.

File: ecommerce/templates/oscar/dashboard/refunds/refund_list.html [hunk 8]
```diff
@@ -7,23 +8,36 @@
...
+    {% endcompress %}
+
+    {% compress js %}
+    <script src="{% static 'vendor/js/jquery.cookie-1.4.1.min.js' %}" type="text/javascript"></script>
+    <script src="{% static 'oscar/js/refund_list.js' %}" type="text/javascript"></script>
+    {% endcompress %}
+{% endblock extrascripts %}
 
 {% block breadcrumbs %}
-    <ul class="breadcrumb">
```
Comment (clintonb) [inline ecommerce/templates/oscar/dashboard/refunds/refund_list.html:14]: Please push the spacing changes to a separate PR.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py (hunk 2 above)
Comment (rlucioni) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: Updated. I've split the tests up to make things clearer.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py (hunk 3 above)
Comment (rlucioni) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: Updated.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py (hunk 4 above)
Comment (rlucioni) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: Updated.

File: ecommerce/static/oscar/js/refund_list.js (hunk 6 above)
Comment (rlucioni) [inline ecommerce/static/oscar/js/refund_list.js]: Updated.

File: ecommerce/static/oscar/js/refund_list.js (hunk 5 above)
Comment (rlucioni) [inline ecommerce/static/oscar/js/refund_list.js]: Good catch, thank you. Updated.

File: ecommerce/templates/oscar/dashboard/orders/order_list.html (hunk 7 above)
Comment (rlucioni) [inline ecommerce/templates/oscar/dashboard/orders/order_list.html:3]: Reverted.

File: ecommerce/templates/oscar/dashboard/refunds/refund_list.html (hunk 8 above)
Comment (rlucioni) [inline ecommerce/templates/oscar/dashboard/refunds/refund_list.html:14]: Reverted. :cry:

**Comment by clintonb** [comment]
:+1:

File: ecommerce/static/oscar/js/refund_list.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        }).success(function (data) {
+            $('tr[data-refund-id=' + refund_id + '] .refund-status').text(data.status);
+            addMessage('alert-success', 'icon-check-sign', 'Refund ' + refund_id + ' has been processed.');
+            $('[data-action=process-refund]').remove();
+        }).fail(function (jqXHR, textStatus, errorThrown) {
+            // NOTE (RFL): For an MVP, changing the displayed refund state on any error may be viable.
+            // Ideally, the displayed refund state would only change if the refund were to enter an
+            // error state. This would be easiest if the processing endpoint returned 200 and a serialized
+            // refund, even when the refund has entered an error state during processing.
+            $('tr[data-refund-id=' + refund_id + '] .refund-status').text('Error');
```
Comment (jimabramson) [inline ecommerce/static/oscar/js/refund_list.js:27]: 'Error' doesn't need translation?

File: ecommerce/static/oscar/js/refund_list.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+        // Make AJAX call and update status
+        $.ajax({
+            url: '/api/v2/refunds/' + refund_id + '/process/',
+            data: { action: decision },
+            method: 'PUT',
+            headers: {'X-CSRFToken': $.cookie('csrftoken')}
+        }).success(function (data) {
+            $('tr[data-refund-id=' + refund_id + '] .refund-status').text(data.status);
+            addMessage('alert-success', 'icon-check-sign', 'Refund ' + refund_id + ' has been processed.');
```
Comment (jimabramson) [inline ecommerce/static/oscar/js/refund_list.js:20]: no translation needed?

File: ecommerce/templates/oscar/dashboard/refunds/refund_detail.html [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+            <th>{% trans "Total Credit" %}</th>
+            <th>{% trans "Created" %}</th>
+            <th>{% trans "Status" %}</th>
+            <th>{% trans "Associated Order" %}</th>
+            <th>{% trans "Actions" %}</th>
+        </tr>
+        <tr data-refund-id="{{ refund.id }}">
+            <td>{{ refund.total_credit_excl_tax|currency:refund.currency }}</td>
+            <td>{{ refund.created }}</td>
+            <td class="refund-status">{{ refund.status|default:"N/A" }}</td>
```
Comment (jimabramson) [inline ecommerce/templates/oscar/dashboard/refunds/refund_detail.html]: curious why the absence of status needs to be handled (also, does "N/A" merit translation?)

File: ecommerce/templates/oscar/dashboard/refunds/refund_detail.html [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+                    <td>{{ line.quantity }}</td>
+                    <td>
+                        {% if line.order_line.product %}
+                        <a href="{% url 'dashboard:catalogue-product' pk=line.order_line.product.id %}">{{ line.order_line.title }}</a>
+                        {% else %}
+                        {{ line.order_line.title }}
+                        {% endif %}
+                    </td>
+                    <td>{{ line.order_line.upc|default:"-" }}</td>
+                    <td>{{ line.status|default:"-" }}</td>
```
Comment (jimabramson) [inline ecommerce/templates/oscar/dashboard/refunds/refund_detail.html]: again, curious about the necessity of default values.

File: ecommerce/templates/oscar/dashboard/refunds/refund_detail.html (hunk 11 above)
Comment (clintonb) [inline ecommerce/templates/oscar/dashboard/refunds/refund_detail.html]: `status` is not nullable, so this is unnecessary.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        """Log into the service and navigate to the refund list view."""
+        self.selenium.get(self.live_server_url + reverse('auto_auth'))
+        self.selenium.get(self.live_server_url + self.path)
+
+    def _decide(self, approve):
+        """Click the Approve or Deny button and wait for the AJAX call to finish."""
+        if approve:
+            button = self.selenium.find_element_by_css_selector(self.approve_button_selector)
+        else:
+            button = self.selenium.find_element_by_css_selector(self.deny_button_selector)
```
Comment (jimabramson) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: very minor: use the `if` switch just to obtain the selector, don't duplicate the method call.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+        self.assertEqual(unicode(self.refund.id), button.get_attribute('data-refund-id'))
```
Comment (jimabramson) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: this seems redundant, given lines 35+36

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+        WebDriverWait(self.selenium, 0.1).until(lambda d: d.find_element_by_css_selector('#messages .alert'))
+
+    def assert_alert_displayed(self, alert_class, text):
+        """Verifies that the most recent alert has the given class and message."""
+        alert = self.selenium.find_elements_by_css_selector('#messages .alert')[-1]
+        classes = alert.get_attribute('class').split(' ')
+        self.assertIn(alert_class, classes)
+        self.assertEqual(alert.find_element_by_css_selector('.message').text, text)
+
+    @skip("Requires refund processing endpoint.")
```
Comment (jimabramson) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py:62]: do you think it's necessary to add a ticket number or some other reminder to remove the skip?  or do you plan to extract that before merging?

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+
+        # Verify that an alert is displayed.
+        self.assert_alert_displayed('alert-success', 'Refund {} has been processed.'.format(self.refund.id))
+
+    @ddt.data(True, False)
+    def test_processing_failure(self, approve):
+        """
+        Verify behavior when refund processing fails.
+
+        The refund list should display "Approve" and "Deny buttons next to each refund. Clicking either button
```
Comment (jimabramson) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: missing close doublequote after Deny

File: ecommerce/templates/oscar/dashboard/refunds/refund_detail.html (hunk 11 above)
Comment (rlucioni) [inline ecommerce/templates/oscar/dashboard/refunds/refund_detail.html]: Updated.

File: ecommerce/templates/oscar/dashboard/refunds/refund_detail.html (hunk 12 above)
Comment (rlucioni) [inline ecommerce/templates/oscar/dashboard/refunds/refund_detail.html]: Updated.

**Comment by jimabramson** [comment]
@rlucioni done reviewing - no major concerns.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py (hunk 13 above)
Comment (rlucioni) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: Updated.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py (hunk 14 above)
Comment (rlucioni) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: It does; updated.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py (hunk 15 above)
Comment (rlucioni) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py:62]: I think the skips are pretty clear when the tests are run (4 capital S's), but I've created [XCOM-342](https://openedx.atlassian.net/browse/XCOM-342) to make sure we remember to remove them.

File: ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py (hunk 16 above)
Comment (rlucioni) [inline ecommerce/extensions/dashboard/refunds/tests/test_acceptance.py]: Updated.

File: ecommerce/static/oscar/js/refund_list.js (hunk 9 above)
Comment (rlucioni) [inline ecommerce/static/oscar/js/refund_list.js:27]: From offline discussion: there's some uncertainty about how best to mark strings in JS for translation. Since we're not worrying about localization for the Otto management interface right now, I've filed [XCOM-343](https://openedx.atlassian.net/browse/XCOM-343) to address this issue later.

File: ecommerce/static/oscar/js/refund_list.js (hunk 10 above)
Comment (rlucioni) [inline ecommerce/static/oscar/js/refund_list.js:20]: Captured in [XCOM-343](https://openedx.atlassian.net/browse/XCOM-343).

**Comment by rlucioni** [comment]
@jimabramson I've addressed your comments.

**Comment by jimabramson** [comment]
:+1:
